privacyidea/lib/tokens/pushtoken.py

```diff
@@ -0,0 +1,578 @@
+
+        :param passw: password, which might be pin or pin+otp
+        :param options: dictionary of additional request parameters
+
+        :return: returns true or false
+        """
+        return self.check_pin(passw, user=user, options=options)
+
+    def create_challenge(self, transactionid=None, options=None):
+        """
+        This method creates a challenge, which is submitted to the user.
+        The submitted challenge will be preserved in the challenge
+        database.
+
+        If no transaction id is given, the system will create a transaction
+        id and return it, so that the response can refer to this transaction.
+
+        :param transactionid: the id of this challenge
+        :param options: the request context parameters / data
+        :type options: dict
+        :return: tuple of (bool, message, transactionid, attributes)
+        :rtype: tuple
+
+        The return tuple builds up like this:
+        ``bool`` if submit was successful;
+        ``message`` which is displayed in the JSON response;
+        additional ``attributes``, which are displayed in the JSON response.
+        """
+        options = options or {}
+        message = get_action_values_from_options(SCOPE.AUTH,
+                                                 ACTION.CHALLENGETEXT,
+                                                 options) or DEFAULT_CHALLENGE_TEXT
+
+        attributes = None
+        data = None
+        challenge = b32encode(geturandom())
+        fb_identifier = self.get_tokeninfo(PUSH_ACTION.FIREBASE_CONFIG)
+        if fb_identifier:
+            # We send the challenge to the Firebase service
+            fb_gateway = create_sms_instance(fb_identifier)
+            url = fb_gateway.smsgateway.option_dict.get(FIREBASE_CONFIG.REGISTRATION_URL)
+            message_on_mobile = get_action_values_from_options(SCOPE.AUTH,
+                                                   PUSH_ACTION.MOBILE_TITLE,
+                                                   options) or DEFAULT_MOBILE_TEXT
+            title = get_action_values_from_options(SCOPE.AUTH,
+                                                   PUSH_ACTION.MOBILE_TITLE,
+                                                   options) or "privacyIDEA"
+            smartphone_data = {"nonce": challenge,
+                               "question": message_on_mobile,
+                               "serial": self.token.serial,
+                               "title": title,
+                               "url": url}
+            # Create the signature.
+            # value to string
+            sign_string = u"{nonce}|{url}|{serial}|{question}|{title}".format(**smartphone_data)
+
+            pem_privkey = self.get_tokeninfo(PRIVATE_KEY_SERVER)
+            privkey_obj = serialization.load_pem_private_key(str(pem_privkey), None, default_backend())
+
+            # Sign the data with PKCS1 padding. Not all Androids support PSS padding.
```

Comment:
I have some thoughts about the choice of RSA signatures with PKCS#1 v1.5 padding:
While PKCS#1 v1.5 signatures are not broken ([see here](https://crypto.stackexchange.com/questions/51547/security-of-rsa-signature-using-pkcs-1-5)), the general recommendation for new applications seems to be PSS padding ([see here](https://cryptography.io/en/latest/hazmat/primitives/asymmetric/rsa/#cryptography.hazmat.primitives.asymmetric.padding.PKCS1v15)).

Instead of RSA, one could even use some elliptic curve signature scheme like ed25519 (recommended by [cryptography](https://cryptography.io/en/latest/hazmat/primitives/asymmetric/ed25519/)). In our case, using ed25519 instead of RSA would reduce the size of stored keys (2048 bits to only 256 bits).

The comment suggests that we cannot use PSS because of legacy(?) Androids. I guess ed25519 isn't better in this regard :-)

But what do you think about adding a "signature" field to the ``smartphone_data``, which currently says "rsa_sha256_pkcs1v15"? This would allow us to upgrade the signature scheme without too much hassle at some point in the future.

Reply:
After a good nights sleep, I realized that ``smartphone_data`` isn't the right place to add the signature algorithm field, it would need to be part of the enrollment process.

Reply:
I like it. But I am not sure, if this should stop our current going forward.

Reply:
I agree that this would mean some extra work now. But including a version marker at some place now may save us time and headaches in the future when we want to migrate to some other signature algorithm.

Reply:
will do

Reply:
We will add "v=1" to the QR Code and the smartphone will check, that it is not v=2.

Reply:
done in 21b07b3f